src/sidebar/components/tag-editor.js

```diff
@@ -1,35 +1,201 @@
+  const inputEl = useRef(null);
+  const [suggestions, setSuggestions] = useState([]);
+
+  /**
+   * Handle changes to this annotation's tags
+   */
+  const updateTags = tagList => {
+    // update suggested-tags list via service
+    tagsService.store(tagList.map(tag => ({ text: tag })));
+    // invoke callback with updated tags
+    onEditTags({ tags: tagList });
   };
 
-  this.autocomplete = function(query) {
-    return Promise.resolve(tags.filter(query));
+  /**
+   * Remove a tag from this annotation.
+   *
+   * @param {string} tag
+   */
+  const removeTag = tag => {
+    const newTagList = [...tagList]; // make a copy
+    const index = newTagList.indexOf(tag);
+    newTagList.splice(index, 1);
+    updateTags(newTagList);
   };
 
-  this.$onChanges = function(changes) {
-    if (changes.tags) {
-      this.tagList = changes.tags.currentValue.map(function(tag) {
-        return { text: tag };
+  /**
+   * Adds a tag equal to the value of the input field, and
+   * then clears out the suggestions list and the input field.
+   */
+  const addTag = () => {
+    const value = inputEl.current.value.trim();
+    if (value.length === 0) {
+      // don't add an empty tag
+      return;
+    }
+    if (tagList.indexOf(value) >= 0) {
+      // don't add duplicate tag
+      return;
+    }
+    updateTags([...tagList, value]);
+    setSuggestions([]);
+
+    // clear the input field and maintain focus
+    inputEl.current.value = '';
+    inputEl.current.focus();
+  };
+
+  /**
+   *  If the user pressed enter or comma while focused on
+   *  the <input>, then add a new tag.
+   */
+  const handleKeyPress = e => {
+    if (e.key === 'Enter' || e.key === ',') {
+      addTag();
+      // preventDefault stops the delimiter from being
+      // added to the input field
+      e.preventDefault();
+    }
+  };
+  /**
+   *  As the user types, change the suggestions query
+   *  based on the value of inputEl.
+   */
+  const handleKeyUp = e => {
+    // Helper to remove any repeated suggestions
+    // that are already tags.
+    const removeDuplicates = suggestions => {
+      const suggestionsSet = [];
+      suggestions.forEach(suggestion => {
+        if (tagList.indexOf(suggestion) < 0) {
+          suggestionsSet.push(suggestion);
+        }
       });
+      return suggestionsSet;
+    };
+
+    if (e.key) {
+      // A user typed a letter and we need to filter the results
+      // to update the <dataset> suggestions
+
+      const value = inputEl.current.value.trim();
+      if (value.length !== 0) {
+        // Only change suggestions if input is non-empty
+        const suggestions = removeDuplicates(tagsService.filter(value));
+        setSuggestions(suggestions);
+      } else {
+        setSuggestions([]);
+      }
+    } else if (e.type === 'keyup') {
+      // Assume the user selected an option from <datalist>
+      // either by clicking on it with the mouse or pressing
```

Comment:
I see... onInput() handles both an `Event` type an `InputEvent` type. Only one has the '.inputType' prop! ahh such a face palm, but I'm glad you discovered that as I didn't notice that before.  This is much better and almost an exact replacement for my key up handler, but after implementing this I found a shortcoming. If you happen to type out the exact tag that is also a suggestion then there could be a problem.

![tabs5](https://user-images.githubusercontent.com/3939074/70786386-6acb2780-1d41-11ea-9b05-dc707cfc0fe7.gif)

When you type out a tag, you have 4 ways of adding that tag.

1. press Enter w/o selecting the option
2. press comma
3. press down arrow once to highlight the item in the options, then press Enter
4. click the option with mouse.

Only case 1 and 2 work. Both 3 and 4 fail because onInput() is never called. Only the keyup even fires in this case. Now I can try to tease out this exact case only for keyUp, but its messy with so many event handlers and still think there are differences between the browsers. I know that the keyup event in FF reports a key "Enter" when you press enter on the datalist, but not in Chrome, for example. I'll have to flush that out a bit. I have a new diff, but too tried to post  atm, I will sync up with you in the morning.